clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -252,17 +252,51 @@ std::set<ConstraintVariable *> ConstraintResolver::getExprConstraintVars(
       switch (UO->getOpcode()) {
       // &e
       case UO_AddrOf: {
-        // We are taking the address, so don't assign to the LHS
         T = getExprConstraintVars(UOExpr, UOExpr->getType());
         std::set<ConstraintVariable *> tmp;
-        /* BUG: Shouldn't always be PTR; could be WILD, so make constraint */
         if (T.empty()) { // doing &x where x is a non-pointer
-          tmp.insert(PVConstraint::getPtrPVConstraint(Info.getConstraints()));
+          UOExpr = UOExpr->IgnoreParenImpCasts();
+          UOExpr = getNormalizedExpr(UOExpr);
+
+          UnaryOperator *SubUO = dyn_cast<UnaryOperator>(UOExpr);
+          if (SubUO && SubUO->getOpcode() == UO_Deref) {
+            // Taking the address of a dereference is a NoOp, so the constraint
+            // vars for the subexpression can be passed through.
```

Comment:
This kind of logic (that `&` annihilates `*`) is needed even with `T` is not empty, right? Why only here and not the `else` case?